core/ibft.go

```diff
@@ -0,0 +1,1134 @@
+			NumMessages: 1,
+			HasMinRound: true,
+		})
+	)
+
+	defer i.messages.Unsubscribe(sub.ID)
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case round := <-sub.SubCh:
+			rcc := i.handleRoundChangeMessage(
+				&proto.View{
+					Height: height,
+					Round:  round,
+				},
+				quorum,
+			)
+			if rcc == nil {
+				continue
+			}
+
+			//	we received a valid RCC for a higher round
+			i.signalNewRCC(ctx, round)
+
+			return
+		}
+	}
+}
+
+// RunSequence runs the IBFT sequence for the specified height
+func (i *IBFT) RunSequence(ctx context.Context, h uint64) {
+	// Set the starting state data
+	i.state.clear(h)
+	i.messages.PruneByHeight(h)
+
+	i.log.Info("sequence started", "num", h)
+	defer i.log.Info("sequence done", "num", h)
+
+	for {
+		i.wg.Add(4)
+
+		view := i.state.getView()
+
+		i.log.Info("round started", "num", view.Round)
+
+		currentRound := view.Round
+		ctxRound, cancelRound := context.WithCancel(ctx)
+
+		// Start the round timer worker
+		go i.startRoundTimer(ctxRound, currentRound)
+
+		//	Jump round on proposals from higher rounds
+		go i.watchForFutureProposal(ctxRound)
+
+		//	Jump round on certificates
+		go i.watchForRoundChangeCertificates(ctxRound)
+
+		// Start the state machine worker
+		go i.runRound(ctxRound)
+
+		teardown := func() {
+			cancelRound()
+			i.wg.Wait()
+		}
+
+		select {
+		case ev := <-i.newProposal:
+			teardown()
+			i.log.Info("received future proposal", "round", ev.round)
+
+			i.moveToNewRound(ev.round)
+			i.acceptProposal(ev.proposalMessage)
+			i.state.setRoundStarted(true)
+		case round := <-i.roundCertificate:
+			teardown()
+			i.log.Info("received future RCC", "round", round)
+
+			i.moveToNewRound(round)
+		case <-i.roundExpired:
+			teardown()
+			i.log.Info("round timeout expired", "round", currentRound)
+
+			newRound := currentRound + 1
+			i.moveToNewRound(newRound)
+
+			i.transport.Multicast(
+				i.backend.BuildRoundChangeMessage(
+					i.state.getLatestPreparedProposedBlock(),
+					i.state.getLatestPC(),
+					&proto.View{
+						Height: h,
+						Round:  newRound,
+					},
+				),
+			)
+		case <-i.roundDone:
+			// The consensus cycle for the block height is finished.
+			// Stop all running worker threads
+			teardown()
+
+			return
+		case <-ctx.Done():
+			teardown()
+			i.log.Debug("sequence cancelled")
+
+			return
+		}
+	}
+}
+
+// runRound runs the state machine loop for the current round
+func (i *IBFT) runRound(ctx context.Context) {
+	// Register this worker thread with the barrier
+	defer i.wg.Done()
+
+	if !i.state.isRoundStarted() {
+		// Round is not yet started, kick the round off
+		i.state.changeState(newRound)
+		i.state.setRoundStarted(true)
+	}
+
+	var (
+		id   = i.backend.ID()
+		view = i.state.getView()
+	)
+
+	// Check if any block needs to be proposed
+	if i.backend.IsProposer(id, view.Height, view.Round) {
+		i.log.Info("we are the proposer")
+
+		proposalMessage := i.buildProposal(ctx, view)
+		if proposalMessage == nil {
+			return
```

Comment:
When may `proposalMessage` be nil? I think we need to emit warning log if proposalMessage is nil.